R/grade_this.R

```diff
@@ -269,6 +270,26 @@ NULL
 #' @export
 .user_code <- placeholder(".user_code")
 
+assert_user_code <- function(user_code, env, frame = parent.frame()) {
+  if (is_placeholder(user_code, ".user_code")) {
+    user_code <- get_from_env(".user_code", env)
+    assert_object_found_in_env(user_code, env, "fail_if_code_feedback")
+  }
+
+  if (is.null(user_code) || length(user_code) == 0 || !nzchar(user_code)) {
+    rlang::return_from(
+      frame,
+      graded(
+        logical(),
+        "I didn't receive your code. Did you write any?",
+        type = "info"
+      )
+    )
```

Comment:
Let's use standard returns here and rely on capturing the graded condition if we want to exit early. One idea would be to call `graded()` and then return `""` if bad user code, otherwise return the discovered `user_code`.

In calling functions we can rely on `capture_graded()` or `tryCatch()` or even `ignore_graded()` to choose how to handle the graded condition.